kaali001/varta
# feat: Added Server-Side Logic for WebRTC Video Chat Application

### Description

This PR introduces the server-side logic to support real-time peer-to-peer video chat using Socket.io and WebRTC signaling. The backend acts as a signaling server to manage user connections, room creation, offer/answer exchanges, and ICE candidates. Below is a breakdown of the changes introduced.


### Changes Made

1. `index.js` – Server entry point:

    - Sets up Express and Socket.io for the backend.
    - Manages client connections and disconnections.
    - Initializes the `UserManager` to handle active users.
    - Listens for WebRTC signaling events (offer, answer, add-ice-candidate).


2. `RoomManager.js` – Manages WebRTC rooms:

   - Creates rooms dynamically when two users are available.
   - Stores room details, including connected users.
   - Facilitates offer and answer exchanges between peers.
   - Manages ICE candidate forwarding between participants.


3. `UserManager.js` – Handles user sessions and queuing:

    - Tracks connected users and manages user queues.
    - Automatically pairs users when two are available.
    - Provides methods to add or remove users on connection events.
    - Initializes WebRTC signaling handlers for each user.


### How to Test This PR

1. Setup:

- Run `npm install` to install dependencies.
- Start the server with `node src/index.js`.

2. Open two browser windows/tabs:

- Connect from both windows to trigger room creation.
- Verify that offers, answers, and ICE candidates are exchanged correctly through the server.

3. Expected Logs:

 > "User connected: <socket_id>"
 > "Offer sent from <socket_id> to <other_socket_id>"
 > "Answer sent from <socket_id> to <other_socket_id>"
 > "ICE candidate sent from <socket_id> to <other_socket_id>"

### **Known Issues / Improvements**

- Room cleanup on disconnection can be improved to handle edge cases.
- Adding error handling for failed WebRTC connections.
- Implement session management to persist user data across reconnects (future enhancement).
